### ISSUE TITLE
Implement Autoregressive Training Loop

### ISSUE THREAD

**Issue by timlawrenz** [timlawrenz/jubilant-palm-tree#117]
**Description**: Create a new training script (`train_autoregressive.py`) for the new decoder.

#### Training Strategy: Teacher Forcing

Teacher forcing stabilizes autoregressive training by using ground truth sequences:

```python
def train_step(model, batch, criterion, optimizer):
    """Single training step with teacher forcing"""
    
    text_embeddings = batch['text_embeddings']
    sequence_length = batch['sequence_length']
    
    total_loss = 0
    hidden_state = None
    
    # Iterate through sequence steps
    for step in range(sequence_length):
        # Ground truth partial graph up to step i
        partial_graph = batch['partial_graphs'][step]
        
        # Ground truth target for step i+1
        target_node_type = batch['target_node_types'][step]
        target_connections = batch['target_connections'][step]
        
        # Forward pass
        outputs = model(text_embeddings, partial_graph, hidden_state)
        
        # Calculate loss for this step
        node_type_loss = F.cross_entropy(
            outputs['node_type_logits'], 
            target_node_type
        )
        connection_loss = F.binary_cross_entropy(
            outputs['connection_probs'], 
            target_connections
        )
        
        step_loss = node_type_loss + connection_loss
        total_loss += step_loss
        
        # Update hidden state for next step
        hidden_state = outputs['hidden_state'].detach()
    
    # Average loss across sequence
    avg_loss = total_loss / sequence_length
    
    # Backpropagation
    optimizer.zero_grad()
    avg_loss.backward()
    optimizer.step()
    
    return avg_loss.item()
```

#### Training Loop Implementation
```python
# train_autoregressive.py
import torch
from torch.utils.data import DataLoader
from src.models import AutoregressiveASTDecoder
from src.data_processing import AutoregressiveASTDataset

def train_autoregressive_decoder():
    # Initialize model
    model = AutoregressiveASTDecoder(
        text_embedding_dim=64,
        graph_hidden_dim=64,
        state_hidden_dim=128,
        node_types=74
    )
    
    # Load autoregressive dataset
    train_dataset = AutoregressiveASTDataset("dataset/train_paired_data.jsonl")
    val_dataset = AutoregressiveASTDataset("dataset/validation_paired_data.jsonl")
    
    train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=8, shuffle=False)
    
    # Training configuration
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=5)
    
    best_val_loss = float('inf')
    patience_counter = 0
    
    for epoch in range(50):
        # Training phase
        model.train()
        train_losses = []
        
        for batch in train_loader:
            loss = train_step(model, batch, optimizer)
            train_losses.append(loss)
        
        # Validation phase
        model.eval()
        val_losses = []
        
        with torch.no_grad():
            for batch in val_loader:
                val_loss = validate_step(model, batch)
                val_losses.append(val_loss)
        
        # Early stopping and checkpointing
        avg_val_loss = sum(val_losses) / len(val_losses)
        
        if avg_val_loss < best_val_loss:
            best_val_loss = avg_val_loss
            torch.save({
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'epoch': epoch,
                'val_loss': avg_val_loss
            }, 'best_autoregressive_decoder.pt')
            patience_counter = 0
        else:
            patience_counter += 1
            
        if patience_counter >= 10:
            print(f"Early stopping at epoch {epoch}")
            break
        
        scheduler.step(avg_val_loss)
```

#### Definition of Done
- [ ] Training script successfully trains autoregressive decoder
- [ ] Teacher forcing used during training for stability
- [ ] Loss calculated at each generation step and averaged
- [ ] Early stopping and learning rate scheduling implemented
- [ ] Best model weights saved based on validation performance

The full description of phase 7 can be found in README_phase7.md, which needs updating to meet the state of this ticket.
